lib/endpoint.js

```diff
@@ -90,17 +90,89 @@ VertoEndpoint.prototype.attemptInvite = function(vertoCall, matrixSide, force) {
     if (matrixSide.candidates.length === 0) { return Promise.resolve(); }
     var self = this;
 
-    var enoughCandidates = false;
+    // see if we've got enough candidates to detrickle correctly.
+    // we start off by gathering some stats on the candidates we have:
+    // splitting into media sections, do we have host/srflx/relay candidates yet?
+    // and sum the component IDs for each foundation (which should sum to 1 or 3
+    // depending on whether we've received just RTP or RTP+RTCP)
+    var stats = [];
+
+    // create empty stats blocks where we expect to see candidates so we can
+    // spot if candidates are entirely missing...
+    for (var i = 0; i < matrixSide.offer.match(/^m=/mg).length; i++) {
+        stats[i] = {};
+    }
+
     for (var i = 0; i < matrixSide.candidates.length; i++) {
         var c = matrixSide.candidates[i];
         if (!c.candidate) { continue; }
-        // got enough candidates when SDP has a srflx or relay candidate
-        if (c.candidate.indexOf("typ srflx") !== -1 ||
-                c.candidate.indexOf("typ relay") !== -1) {
-            enoughCandidates = true;
-            console.log("Gathered enough candidates for %s", matrixSide.mxCallId);
-            break; // bail early
+
+        var mLineIndex;
+        if (c.sdpMLineIndex !== undefined) {
+            mLineIndex = c.sdpMLineIndex;
+        }
+        else {
+            // hack: synthesise an mLineIndex based on heuristics
+            if (c.sdpMid === 'audio') {
+                mLineIndex = 0;
+            }
+            else if (c.sdpMid === 'video') {
+                mLineIndex = 1;
+            }
+            else {
+                mLineIndex = 0; // this is probably wrong though.
+            }
+        }
+        stats[mLineIndex] = stats[mLineIndex] || {};
+        if (c.candidate.indexOf("typ host") !== -1) {
+            stats[mLineIndex].hasHost = 1;
+        }
+        if (c.candidate.indexOf("typ srflx") !== -1) {
+            stats[mLineIndex].hasSrflx = 1;
+        }
+        if (c.candidate.indexOf("typ relay") !== -1) {
+            stats[mLineIndex].hasRelay = 1;   
         }
+        var m = c.candidate.match(/^candidate:(.*?) (.*?) /);
+        if (m) {
+            stats[mLineIndex].componentIdSumByFoundation =
+                stats[mLineIndex].componentIdSumByFoundation || {};
+            stats[mLineIndex].componentIdSumByFoundation[m[1]] += m[2];
```

Comment:
You mention above that:

> sum the component IDs for each foundation (which should sum to 1 or 3 depending on whether we've received just RTP or RTP+RTCP)

which I'm guessing is what this is doing. I would really like to see an example comment and ideally some better var names than `m[1]` (though I appreciate this is a result of a regex match). Some sanity checks to make sure the matched groups can be parsed as `Numbers` would be good too!